snuba/query/joins/classifier.py

```diff
@@ -0,0 +1,250 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, replace
+from functools import partial
+from typing import (
+    Callable,
+    List,
+    Mapping,
+    MutableMapping,
+    NoReturn,
+    Optional,
+    Sequence,
+    Set,
+)
+
+from snuba.query.expressions import (
+    Argument,
+    Column,
+    CurriedFunctionCall,
+    Expression,
+    ExpressionVisitor,
+    FunctionCall,
+    Lambda,
+    Literal,
+    SubscriptableReference,
+)
+
+
+def assert_never(value: NoReturn) -> NoReturn:
+    assert False, f"Unhandled value: {value} ({type(value).__name__})"
+
+
+@dataclass(frozen=True)
+class SubExpression(ABC):
+    """
+    Data structure maintained when visiting an Expression in the query.
+    This keeps track of the main branch (the node we are visiting) as
+    well as all the branches we cut that were under the main expression.
+    Those cut branches should be pushed down into subqueries.
+    """
+
+    main_expression: Expression
```

Comment:
Should this be optional? I ask because there are some expressions you want to push down that don't need to be referenced in the main query at all, e.g. where conditions. In that case we want to push the expression down but we don't need a main_expression.

Reply:
IT is true that some expressions won't be referenced, though that will be inferred by the code that visits the main expression (that is the one who knows what does need to be referenced and what does not). 
The branch cutter only has knowledge of an individual expression tree. It does not know whether the expression is in the select, where, etc. So it cannot decide when to populate the main_expression and when not to doo so.
In practice when the main query code instantiates the BranchCutter to visit one root level expression, it will decide whether to add a reference in the main query depending whether we are traversing a condition or a group by etc. and based on what is returned by the Subexpression.cut_branch().